File: plugins/logistics/backend/dto/load_plans.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class LoadPlanItemUpsert(BaseModel):
    product_id: str
    planned_quantity: float
    source_warehouse_id: str | None = None
    notes: str | None = Field(default=None, max_length=500)


class LoadPlanUpsertRequest(BaseModel):
    notes: str | None = Field(default=None, max_length=500)
    items: list[LoadPlanItemUpsert]


class ConfirmLoadRequest(BaseModel):
    notes: str | None = Field(default=None, max_length=500)


class ReturnRemainingRequest(BaseModel):
    destination_warehouse_id: str | None = None
    notes: str | None = Field(default=None, max_length=500)


class LoadPlanItemRead(BaseModel):
    id: str
    product_id: str
    product_name: str
    planned_quantity: float
    planned_weight_kg: float | None = None
    source_warehouse_id: str
    notes: str | None = None
    requires_serials: bool = False
    selected_serials_count: int = 0
    serials_complete: bool = True
    created_at: datetime


class LoadPlanRead(BaseModel):
    id: str | None = None
    session_id: str
    status: str
    notes: str | None = None
    planned_weight_kg: float = 0
    items: list[LoadPlanItemRead] = Field(default_factory=list)

File: plugins/logistics/frontend/api/load-plans.ts
import { API_PREFIX } from "./_shared";
import { apiRequest } from "@systutor/shell/api/client";

export type LoadPlanItem = {
  id: string;
  product_id: string;
  product_name: string;
  planned_quantity: number;
  planned_weight_kg: number | null;
  source_warehouse_id: string;
  notes: string | null;
  requires_serials: boolean;
  selected_serials_count: number;
  serials_complete: boolean;
  created_at: string;
};

export type LoadPlan = {
  id: string | null;
  session_id: string;
  status: string;
  notes: string | null;
  planned_weight_kg: number;
  items: LoadPlanItem[];
};

export type LoadPlanUpsertPayload = {
  notes?: string | null;
    items: Array<{
      product_id: string;
      planned_quantity: number;
      source_warehouse_id?: string | null;
      notes?: string | null;
  }>;
};

export type ConfirmLoadPayload = {
  notes?: string | null;
};

export type ReturnRemainingPayload = {
  destination_warehouse_id?: string | null;
  notes?: string | null;
};

export function getLoadPlan(sessionId: string) {
  return apiRequest<LoadPlan>(`${API_PREFIX}/vehicle-sessions/${sessionId}/load-plan`);
}

export function upsertLoadPlan(sessionId: string, payload: LoadPlanUpsertPayload) {
  return apiRequest<LoadPlan>(`${API_PREFIX}/vehicle-sessions/${sessionId}/load-plan`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function confirmLoad(sessionId: string, payload: ConfirmLoadPayload = {}) {
  return apiRequest<{ session_id: string; loaded_weight_kg: number }>(
    `${API_PREFIX}/vehicle-sessions/${sessionId}/confirm-load`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    }
  );
}

export function confirmAndReady(sessionId: string, payload: ConfirmLoadPayload = {}) {
  return apiRequest<{ id: string; status: string; loaded_weight_kg: number | null }>(
    `${API_PREFIX}/vehicle-sessions/${sessionId}/confirm-and-ready`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    }
  );
}

export function returnRemaining(sessionId: string, payload: ReturnRemainingPayload = {}) {
  return apiRequest<{ session_id: string; status: string }>(
    `${API_PREFIX}/vehicle-sessions/${sessionId}/return-remaining`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    }
  );
}
